### PR TITLE
Expose ReferenceElement for qualified input components

### PR DESCRIPTION
**Pull request by uabarahona**
# Summary
As discussed on #24446 this PR is to expose "InputElement" which is a `ReferenceElement`, some notes to take into account:

* @MackinnonBuck  I have stick with `InputElement` name everywhere to keep consistency. [I think it's fine as `select` is also a input, but I can change the name for select if it is really needed]
* I have made the `set` protected to support inheritance

Addresses #24446

### CODE REVIEW

File: src/Components/Web/src/Forms/InputCheckbox.cs [hunk 1]
```diff
@@ -21,6 +21,12 @@ namespace Microsoft.AspNetCore.Components.Forms
     /// </summary>
     public class InputCheckbox : InputBase<bool>
     {
+        /// <summary>
+        /// Gets or sets the associated <see cref="ElementReference"/>
+        /// <para>This value is used to gain access to helper methods on the input elment.</para>
+        /// </summary>
+        public ElementReference? InputElement { get; protected set; }
```
Comment (MackinnonBuck) [inline src/Components/Web/src/Forms/InputCheckbox.cs]:
```suggestion
        /// <summary>
        /// Gets or sets the associated <see cref="ElementReference"/>.
        /// </summary>
        public ElementReference? InputElement { get; protected set; }
```

I think this is enough - the uses of this property aren't limited to accessing helper methods (you could pass the `ElementReference` to JS, etc.).

Same applies to the other files.

File: src/Components/Web/src/Forms/InputCheckbox.cs [hunk 2]
```diff
@@ -21,6 +21,12 @@ namespace Microsoft.AspNetCore.Components.Forms
     /// </summary>
     public class InputCheckbox : InputBase<bool>
     {
+        /// <summary>
+        /// Gets or sets the associated <see cref="ElementReference"/>
+        /// <para>This value is used to gain access to helper methods on the input elment.</para>
```
Comment (meziantou) [inline src/Components/Web/src/Forms/InputCheckbox.cs:26]:
```suggestion
        /// <para>This value is used to gain access to helper methods on the input element.</para>
```

Comment (meziantou) [reply src/Components/Web/src/Forms/InputCheckbox.cs:26]: Same typo in other files

File: src/Components/Web/src/Forms/InputSelect.cs [hunk 3]
```diff
@@ -16,6 +16,12 @@ public class InputSelect<TValue> : InputBase<TValue>
         /// </summary>
         [Parameter] public RenderFragment? ChildContent { get; set; }
 
+        /// <summary>
+        /// Gets or sets the associated <see cref="ElementReference"/>
+        /// <para>This value is used to gain access to helper methods on the input elment.</para>
+        /// </summary>
+        public ElementReference? InputElement { get; protected set; }
```
Comment (MackinnonBuck) [inline src/Components/Web/src/Forms/InputSelect.cs:23]: I personally think `SelectElement` would be better as to not give the impression that it's referring to an `<input>` element.

@pranavkm, do you have a strong opinion on this?

File: src/Components/Web/src/Forms/InputCheckbox.cs (hunk 2 above)
Comment (barahonajm) [reply src/Components/Web/src/Forms/InputCheckbox.cs:26]: Thanks! I am going to fix those

**Review by MackinnonBuck — CHANGES_REQUESTED** [review]
Looks good! I had a couple small comments. @pranavkm, could you take a look?

File: src/Components/Web/src/Forms/InputCheckbox.cs (hunk 1 above)
Comment (barahonajm) [reply src/Components/Web/src/Forms/InputCheckbox.cs]: Agree

File: src/Components/Web/src/Forms/InputCheckbox.cs [hunk 4]
```diff
@@ -21,6 +21,11 @@ namespace Microsoft.AspNetCore.Components.Forms
     /// </summary>
     public class InputCheckbox : InputBase<bool>
     {
+        /// <summary>
+        /// Gets or sets the associated <see cref="ElementReference"/>
+        /// </summary>
+        public ElementReference? InputElement { get; protected set; }
```
Comment (pranavkm) [inline src/Components/Web/src/Forms/InputCheckbox.cs:27]: This is a slightly unusual nullable property in that it becomes available and stays non-null once the component is rendered. Would it be better to have this throw if it's accessed before it's initialized?

```C#
    return _inputElement ?? throw new InvalidOperationException($"Component must be rendered before {nameof(InputElement)} can be accessed.");

Comment (pranavkm) [reply src/Components/Web/src/Forms/InputCheckbox.cs:27]: This would also allow the property to be non-null.

Comment (barahonajm) [reply src/Components/Web/src/Forms/InputCheckbox.cs]: Agree

File: src/Components/Web/src/Forms/InputSelect.cs (hunk 3 above)
Comment (barahonajm) [reply src/Components/Web/src/Forms/InputSelect.cs]: Changed

**Review by pranavkm — COMMENTED** [review]
This looks really good. Any chance we could add one test that verifies this works [end-to-end](https://github.com/dotnet/aspnetcore/tree/master/src/Components/test/E2ETest)? Something that verifies that you could Focus on the input element using this mechanism?

**Comment by barahonajm** [comment]
Tests were added and they brought up an issue where `InputRadio` was being re-render always no matter what, this due `InputRadioGroup` sets `InputRadioContext` as key but `InputRadioContext` does not have a way to compare so the diff process was taking the `InputRadioGroups` always as diferent :

https://github.com/dotnet/aspnetcore/blob/91776832e08d5ddf8abc2824ee7ac24f42f953e7/src/Components/Web/src/Forms/InputRadioGroup.cs#L45-L52

**Comment by MackinnonBuck** [comment]
@barahonajm The problem you're describing is intended behavior - the radio buttons are re-rendered when the `_context` changes so validation styling is applied. We might be able to investigate a way to preserve the radio button instances when re-rendering so `ElementReference`s don't get invalidated. Was the issue in using the `ElementReference` itself or just the E2E tests (i.e. `StaleElementReferenceException`)?

**Comment by barahonajm** [comment]
@MackinnonBuck I see, let me ask a little question shouldn't the `radiogroup` be re-renderd only when the `_context` changes? because in my case both `_context`  old and new have exactly the same content.

The issue on E2E test is that the `FocusAsync()` does not work at all for radio buttons since they get re-render after the focus is applied

**Comment by MackinnonBuck** [comment]
@barahonajm it must be because a new `InputRadioContext` is created each time the parameters are set, so it is therefore a new key. One solution could be to get rid of the IsFixed attribute for the CascadingValue and remove the key as well, but I'm not completely sure what the performance implications of this are. @pranavkm, do you have any suggestions for this?

**Comment by barahonajm** [comment]
Yes, that's the reason I was not able to explain it well 😃, I added an equals and that solved the issue, but I don't know too the performance implications

**Comment by pranavkm** [comment]
We discussed options and think it might be easier to not have the `InputRadio.ElementReference` to start with. We can always fix the implementation in the future to have this work.

**Comment by barahonajm** [comment]
Got it, I have removed the Inputradio from the PR

**Comment by mrlife** [comment]
@pranavkm We've seen where if things don't make it into 5.0, they're punted to 6.0. Could this mean 5.x (before November 2021) or does it mean November 2021 at the earliest?

**Review by Toemsel — APPROVED** [review]

**Comment by SteveSandersonMS** [comment]
@pranavkm When completing this PR, please also consider adding support for https://github.com/dotnet/aspnetcore/issues/28533 and closing that as resolved.

**Comment by pranavkm** [comment]
Thanks @JuanElements for your PR. I've created another PR based on your changes here - https://github.com/dotnet/aspnetcore/pull/29555. We'll use that to track this to completion.

**Pull request closed without merging**